### PR TITLE
Update master branch to LVGL v9 version

### PR DESCRIPTION
**Pull request by PGNetHun**
- Update LVGL to v9-rc version, commit: https://github.com/lvgl/lvgl/commit/09cb87cdc6a0168a98bc0a3182a8439b13249ead
- Fix `gen_mpy.py` binding script
- Amend MicroPython files and examples to the new LVGL namings (for example: `img` -> `image`, `btn` -> `button`)
- Update `README.md`: rename `Micropython` -> `MicroPython`
- Remove `lv.COLOR_FORMAT.NATIVE_REVERSED`, and use instead `lv_draw_sw_rgb565_swap()` function for some displays
- Remove `lv.COLOR_DEPTH` from display drivers and examples
- Modify `ili9XXX.py`: change class attributes to instance attributes (-> eliminate "singleton" class attributes) 
- Modify `ili9XXX.py`: set explicit default color formats instead of `lv.COLOR_FORMAT.NATIVE`
- Modify MicroPython display drivers (`ili9XXX.py` and `st77xx.py`) and examples to use new display draw buffer approach (related PR: https://github.com/lvgl/lvgl/pull/5204)
- Update examples to align to LVGL changes
- Disable LVGL log (by default)
- Merge `master` changes into PR branch

### CODE REVIEW

**Comment by PGNetHun** [comment]
@kisvegabor , @XuNeo , @kdschlosser 
Please review this PR before I merge it to `master`.

**Comment by PGNetHun** [comment]
@sam0910 :  this PR contains the latest LVGL with fixed display drivers (including the fix for your feedback: https://github.com/lvgl/lv_binding_micropython/pull/298#issuecomment-1880908299).
Please test this `multi-instance` branch on your device, and if it works, then give an approve to this PR.
Thank you!

**Review by kisvegabor — APPROVED** [review]
I haven't tested it just read the code and as far as I can tell it look good.

**Comment by kisvegabor** [comment]
Waiting for @sam0910's approval to merge.

**Comment by PGNetHun** [comment]
This PR works on M5Core2 device  (ESP32 SPIRAM, ILI9341 display)

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:114]: color_format should not be defaulted to `None`. This is how LVGL now determines how the buffer data is to be stored. It should be defaulted to lv.COLOR_FORMAT.RGB888.

There needs to be a test that gets done so if a user inadvertiantly sets the `swap_rgb565_bytes` flag to true but has a color format other than RGB565 specified the flag will automatically get switched to  `False`

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:560]: do not use this macro as it is no longer being used in LVGL. 

Use `lv.color_format_get_size(self.color_space)` instead

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:610]: `lv.color_format_get_size(self.color_space)`

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:662]: `lv.color_format_get_size(self.color_space)`

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:668]: `lv.color_format_get_size(self.color_space)`

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:691]: `lv.color_format_get_size(self.color_space)`

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:781]: `lv.color_format_get_size(self.color_space)`

File: driver/esp32/ili9XXX.py
Comment (kdschlosser) [inline driver/esp32/ili9XXX.py:833]: `lv.color_format_get_size(self.color_space)`

File: driver/generic/st77xx.py
Comment (kdschlosser) [inline driver/generic/st77xx.py:451]: `lv.color_format_get_size(self.color_space)`

File: driver/esp32/ili9XXX.py
Comment (kisvegabor) [reply driver/esp32/ili9XXX.py:114]: > It should be defaulted to lv.COLOR_FORMAT.RGB888.

RGB565 is more common, so I'd choose it as default.

File: driver/generic/st77xx.py
Comment (kdschlosser) [inline driver/generic/st77xx.py:461]: do not use `lv.COLOR_FORMAT.NATIVE` that is what the `color_format` parameter is for. That is what should be passed when setting the color format

File: driver/esp32/ili9XXX.py
Comment (kisvegabor) [reply driver/esp32/ili9XXX.py:560]: True, now the display can have their own color format set in run time independently of `LV_COLOR_DEPTH`.

File: examples/advanced_demo.py
Comment (kdschlosser) [inline examples/advanced_demo.py:361]: `lv.color_format_get_size(color_format)`

File: examples/advanced_demo.py
Comment (kdschlosser) [inline examples/advanced_demo.py:362]: `lv.color_format_get_size(color_format)`

File: examples/example3.py
Comment (kdschlosser) [inline examples/example3.py:18]: `lv.color_format_get_size(color_format)`

File: examples/example3.py
Comment (kdschlosser) [inline examples/example3.py:19]: `lv.color_format_get_size(color_format)`

**Comment by kdschlosser** [comment]
and would you please separate the display drivers into their own files and import the base class into the file. The amount of memory that gets consumed because of importing a bunch of code for drivers that are not being used is simply out of control.

Also get rid of the class level attribute for the init parameters and put them into a function. they get used only once and to have them sitting there eating up memory after the initialization is just crazy. If the init parameters are set in the \_\_init\_\_ function of the driver you can delete them after they are used. But if they are set at the class level then you cannot delete them because if a user has more than one display attached that is the same the init commands will be gone and will not be able to be used when the second display is initialized.

File: examples/example3.py
Comment (kdschlosser) [reply examples/example3.py:19]: for this one I believe there is a function called `lv.color_format_get_bpp`

File: examples/example3.py
Comment (kdschlosser) [reply examples/example3.py:18]: for this one I believe there is a function called lv.color_format_get_bpp

File: examples/advanced_demo.py
Comment (kdschlosser) [reply examples/advanced_demo.py:362]: for this one I believe there is a function called lv.color_format_get_bpp

File: examples/advanced_demo.py
Comment (kdschlosser) [reply examples/advanced_demo.py:361]: for this one I believe there is a function called lv.color_format_get_bpp

File: examples/example3.py
Comment (kisvegabor) [reply examples/example3.py:18]: See [here](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_color.h#L200-L215).

**Comment by kisvegabor** [comment]
I suggest merging this PR if it's doesn't break anything, and send smaller ones after that.

**Comment by kdschlosser** [comment]
and TY for fixing that error in gen_mpy. I mentioned the problem as you know and it was never corrected.

**Comment by kdschlosser** [comment]
since LVGL internally no longer changes things like structures or unions based on the color depth why is there still a reliance on LV_COLOR_DEPTH??

It sets the NATIVE color format enumerations and it is used to set _LV_COLOR_NATIVE_WITH_ALPHA_SIZE which is used in LV_IMAGE_BUF_SIZE_TRUE_COLOR_ALPHA. This macro is not accessible in the binding nor is it used at all in LVGL's internal code.

the color depth is also used in LV_IMAGE_BUF_SIZE_TRUE_COLOR_CHROMA_KEYED and LV_IMAGE_BUF_SIZE_TRUE_COLOR and none of those macros are used anywhere in LVGL code. Those macros are not exposed in the binding and will never get used. So setting LV_COLOR_DEPTH does not need to be done at all. It's is pointless to do so. It adds additional complexity and it also forces the binding to be a compile time thing instead of a runtime thing. If LV_COLOR_DEPTH is not used at all then a user will not have ro compile and flash their firmware if they want to change out to a different display. All they have to do is upload the new python source file that has the new driver code in it.  That's it. The sole point to using python is rapid development. having to recompile firmware and flash firmware if changing a piece of hardware that is attached kind of negates the point to using it.

LV_COLOR_DEPTH also sets antialiasing in the display driver but that is easily changed in python code.

**Comment by sam0910** [comment]
@kisvegabor, I've just compliled this PR with lv_micropython/feat-multiinstance.
It is working without ANY issue. I am so happy to have the latest LVGL with micropython 1.20.0 :)

**Comment by sam0910** [comment]
with lvgl@a9960c62 is working any issue :)
Thank you so much, and I am very excited that I could use the latest v9,
with great new features.

I have not measured the detail, but the rendering speed is better than v8.
I could feel it :)

**Comment by kisvegabor** [comment]
> I have not measured the detail, but the rendering speed is better than v8.
I could feel it :)

Ahhh, what a relief :slightly_smiling_face:

**Comment by kisvegabor** [comment]
We can remove `LV_IMAGE_BUF_SIZE_` as they are not used anymore. See [5244](https://github.com/lvgl/lvgl/pull/5244)

`LV_COLOR_DEPTH` still affects a few minor places and we have no time to remove it for v9. We can get beck to this in the near future.

File: driver/esp32/espidf.c
Comment (XuNeo) [inline driver/esp32/espidf.c:12]: Is this header really needed?

File: examples/png_example.py
Comment (XuNeo) [inline examples/png_example.py:4]: This is going to be changed again, but current macro name is `LV_CACHE_DEF_SIZE`.
# For dragging to work reasonable, make sure LV_image_CACHE_DEF_SIZE is not 0!
```suggestion
# For dragging to work reasonable, make sure LV_CACHE_DEF_SIZE is not 0!
```

File: driver/esp32/espidf.c
Comment (PGNetHun) [reply driver/esp32/espidf.c:12]: Yes, because we call `lv_draw_sw_rgb565_swap()` function in `ili9xxx_flush()`

Comment (PGNetHun) [reply driver/esp32/espidf.c:12]: Sorry, not needed, `lvgl.h` is already included

File: examples/png_example.py
Comment (PGNetHun) [reply examples/png_example.py]: Fixed. It was replaced accidentally.

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed.
Error messages are also updated.

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed

File: examples/example3.py
Comment (PGNetHun) [reply examples/example3.py]: Fixed.
Use XRGB888 color format for this **rk043fn48h** display

File: examples/example3.py
Comment (PGNetHun) [reply examples/example3.py]: Fixed.

File: examples/advanced_demo.py
Comment (PGNetHun) [reply examples/advanced_demo.py]: Fixed.
Use ARGB888 color format for this **rk043fn48h** display

File: examples/advanced_demo.py
Comment (PGNetHun) [reply examples/advanced_demo.py]: Fixed

File: driver/generic/st77xx.py
Comment (PGNetHun) [reply driver/generic/st77xx.py]: Fixed.
Because this is ST77xx, the color format is explicitly set to RGB565.

File: driver/generic/st77xx.py
Comment (PGNetHun) [reply driver/generic/st77xx.py]: Fixed.

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed.

File: driver/esp32/ili9XXX.py
Comment (PGNetHun) [reply driver/esp32/ili9XXX.py]: Fixed.

**Comment by PGNetHun** [comment]
> and would you please separate the display drivers into their own files and import the base class into the file. The amount of memory that gets consumed because of importing a bunch of code for drivers that are not being used is simply out of control.
> 
> Also get rid of the class level attribute for the init parameters and put them into a function. they get used only once and to have them sitting there eating up memory after the initialization is just crazy. If the init parameters are set in the __init__ function of the driver you can delete them after they are used. But if they are set at the class level then you cannot delete them because if a user has more than one display attached that is the same the init commands will be gone and will not be able to be used when the second display is initialized.

Next step in a new PR.

**Comment by PGNetHun** [comment]
The new changes are tested on ST7789 display (ESP32), and it works. For me the performance (FPS) is worse than with LVGL 8.3, but I will investigate this issue later.

@sam0910 , could you please test again on your device this updated PR? Thank you!

**Comment by sam0910** [comment]
> The new changes are tested on ST7789 display (ESP32), and it works. For me the performance (FPS) is worse than with LVGL 8.3, but I will investigate this issue later.
> 
> @sam0910 , could you please test again on your device this updated PR? Thank you!

[Update ili9xxx, st77xx drivers and examples, remove lv.COLOR_DEPTH](https://github.com/lvgl/lv_binding_micropython/commit/5cea992a704bf66836f32010422c78a31b974577) is looking great! no issues!

it might not be relevant with this PR, but I found a roller widget has 0ms anim_time by default. so user set ```roller.set_selected(1, lv.ANIM.ON)``` they cannot see the anim, unless they set ```roller.set_style_anim_time(1000, lv.PART.MAIN)```

**Comment by kisvegabor** [comment]
Please note these merged PR: https://github.com/lvgl/lvgl/pull/5204
There are no other planned API changes before v9.

**Comment by PGNetHun** [comment]
> Please note this merged PR: [lvgl/lvgl#5204](https://github.com/lvgl/lvgl/pull/5204) There are no other planned API changes before v9.

Get stuck with an issue about allocating ESP32 DMA-capable memory for display buffers:
https://github.com/lvgl/lv_binding_micropython/issues/313

**Comment by PGNetHun** [comment]
> > Please note this merged PR: [lvgl/lvgl#5204](https://github.com/lvgl/lvgl/pull/5204) There are no other planned API changes before v9.
> 
> Get stuck with an issue about allocating ESP32 DMA-capable memory for display buffers: #313

Problem solved in another [PR](https://github.com/lvgl/lv_binding_micropython/pull/314), and it is merged to this `multi-instance` branch.

**Comment by PGNetHun** [comment]
@kisvegabor , @XuNeo , @kdschlosser ,
Could you please review & approve this big PR?
I don't want to add any more commits to it, so it is ready to merge to `master` branch.
After this, every new PRs should target `master`.

(this PR is tested on M5Core2 ESP32 device, and it works)

**Review by kisvegabor — APPROVED** [review]
I'm not super confident, but as far as I can tell it's ok :slightly_smiling_face:

**Comment by PGNetHun** [comment]
Reminder: disable `LV_USE_ARABIC_PERSIAN_CHARS`
see: https://github.com/lvgl/lvgl/pull/5386

File: examples/example1.py
Comment (PGNetHun) [reply examples/example1.py]: Fixed

**Comment by PGNetHun** [comment]
To have a linear and clean git history on master branch, use "Squash and merge" merge strategy.

**Review by kisvegabor — APPROVED** [review]

**Review by XuNeo — APPROVED** [review]
I can only review part of the code. It looks good to me.

**Pull request merged**
